modules/backendscheduler/backendscheduler.go

```diff
@@ -0,0 +1,391 @@
+		if j.Status() == work.JobStatusPending || j.Status() == work.JobStatusRunning {
+			if j.WorkerID() == req.WorkerId {
+				resp := &tempopb.NextJobResponse{
+					JobId:  j.ID,
+					Type:   j.Type,
+					Detail: j.JobDetail,
+				}
+
+				level.Info(log.Logger).Log("msg", "assigned previous job to worker", "job_id", id, "worker", req.WorkerId)
+				return resp, nil
+			}
+		}
+	}
+
+	// Find next available job
+	for id, j := range s.work.Jobs() {
+		if j.Status() == work.JobStatusPending {
+			// Honor the request job type if specified
+			if req.Type != tempopb.JobType_JOB_TYPE_UNSPECIFIED && j.Type != req.Type {
+				continue
+			}
+
+			// Create response with job details
+			resp := &tempopb.NextJobResponse{
+				JobId:  j.ID,
+				Type:   j.Type,
+				Detail: j.JobDetail,
+			}
+
+			// Mark job as running
+			j.Start(req.WorkerId)
+
+			level.Info(log.Logger).Log("msg", "assigned job to worker", "job_id", id, "worker", req.WorkerId)
+			return resp, nil
+		}
+	}
+
+	// No jobs available
+	return &tempopb.NextJobResponse{}, nil
+}
+
+// UpdateJob implements the BackendSchedulerServer interface
+func (s *BackendScheduler) UpdateJob(_ context.Context, req *tempopb.UpdateJobStatusRequest) (*tempopb.UpdateJobStatusResponse, error) {
+	j := s.work.GetJob(req.JobId)
+	if j == nil {
+		return nil, fmt.Errorf("job not found: %s", req.JobId)
+	}
+
+	switch req.Status {
+	case tempopb.JobStatus_JOB_STATUS_SUCCEEDED:
+		j.Complete()
+		metricJobsCompleted.WithLabelValues(j.JobDetail.Tenant, j.JobDetail.Tenant).Inc()
+		level.Info(log.Logger).Log("msg", "job completed", "job_id", req.JobId)
+
+		// TODO: update the blocklist?
+
+	case tempopb.JobStatus_JOB_STATUS_FAILED:
+		j.Fail()
+		metricJobsFailed.WithLabelValues(j.JobDetail.Tenant, j.JobDetail.Tenant).Inc()
+		level.Error(log.Logger).Log("msg", "job failed", "job_id", req.JobId, "error", req.Error)
+
+	default:
+		return nil, fmt.Errorf("invalid job status: %v", req.Status)
+	}
+
+	return &tempopb.UpdateJobStatusResponse{
+		Success: true,
+	}, nil
+}
+
+// ListJobs returns all jobs for a given tenant
+func (s *BackendScheduler) ListJobs(_ context.Context) []*work.Job {
+	return s.work.Jobs()
+}
+
+// CreateCompactionJob creates a new compaction job for the given tenant and blocks.
+// Must be called under jobsMtx lock.
+func (s *BackendScheduler) createCompactionJob(_ context.Context, tenantID string, input []string) error {
+	// Skip blocks which already have a job
+	for _, blockID := range input {
+		for _, j := range s.work.Jobs() {
+			if j.JobDetail.Tenant == tenantID {
+				switch j.Type {
+				case tempopb.JobType_JOB_TYPE_COMPACTION:
+					for _, b := range j.JobDetail.Compaction.Input {
+						if b == blockID {
+							// TODO: consider continue when we have a stripe of like-blocks.
+							return nil
+						}
+					}
+				default:
+					continue
+				}
+			}
+		}
+	}
+
+	jobID := uuid.New().String()
+
+	job := &work.Job{
+		ID:   jobID,
+		Type: tempopb.JobType_JOB_TYPE_COMPACTION,
```

Comment:
specify "pending" or whatever?

Reply:
I'm leveraging "UNSPECIFIED" for this case currently.  The worker will then report the status as RUNNING when it starts the work.